File: ios/SketchView/SketchView.m
#import "SketchView.h"
#import "PenSketchTool.h"
#import "EraserSketchTool.h"
#import "RectangleTool.h"
#import "ArrowTool.h"
#import "TextTool.h"
#import "NSMutableArray+QueueStack.h"
#import "SketchViewContainer.h"


@implementation SketchView
{
    SketchTool *currentTool;
    PenSketchTool *penTool;
    EraserSketchTool *eraseTool;
    RectangleTool *rectangleTool;
    ArrowTool *arrowTool;
    TextTool *textTool;
    
    UIImage *incrementalImage;
    NSMutableArray *stack;
    NSInteger maxUndo;
}

-(instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    [self initialize];
    return self;
}

- (void) initialize
{
    [self setMultipleTouchEnabled:NO];
    maxUndo = 10; // keep it small so we don't overflow our memory
    stack = [NSMutableArray array];
    penTool = [[PenSketchTool alloc] initWithTouchView:self];
    eraseTool = [[EraserSketchTool alloc] initWithTouchView:self];
    rectangleTool = [[RectangleTool alloc] initWithTouchView:self];
    arrowTool = [[ArrowTool alloc] initWithTouchView:self];
    textTool = [[TextTool alloc] initWithTouchView:self];
    
    [self setToolType:SketchToolTypePen];
    
    [self setBackgroundColor:[UIColor clearColor]];
}

-(void)setToolType:(SketchToolType) toolType
{
    if(currentTool){
        [currentTool clear];
        [self setNeedsDisplay];
    }
    
    switch (toolType) {
        case SketchToolTypePen:
            currentTool = penTool;
            break;
        case SketchToolTypeEraser:
            currentTool = eraseTool;
            break;
        case SketchToolTypeRectangle:
            currentTool = rectangleTool;
            break;
        case SketchToolTypeArrow:
            currentTool = arrowTool;
            break;
        case SketchToolTypeText:
            currentTool = textTool;
            break;
        default:
            currentTool = penTool;
            break;
    }
}

-(void)setMaxUndo:(NSInteger)max
{
    maxUndo = max;
    NSUInteger initialSize = [stack count];
    while([stack count] >= maxUndo){
        [stack queuePop];
    }
    
    if(initialSize != [stack count]){
        [self onDrawSketch];
    }
}

// keep pen tool as the source of truth
// but set all that might need use color
-(void)setToolColor:(UIColor *)rgba
{
    [penTool setToolColor:rgba];
    [rectangleTool setToolColor:rgba];
    [arrowTool setToolColor:rgba];
    [textTool setToolColor:rgba];
}

-(UIColor *)getToolColor
{
    return [penTool getToolColor];
}

-(void)setViewImage:(UIImage *)image
{
    if(incrementalImage != nil){
        // if stack full, remove oldest
        if([stack count] >= maxUndo){
            [stack queuePop];
        }
        
        [stack queuePush:incrementalImage];
    }
    // if prev image is null,
    // add a dummy stack element so we can undo it too
    else{
        [stack queuePush:[NSNull null]];
    }
    
    incrementalImage = image;
    [self setNeedsDisplay];
    [self onDrawSketch];
}

-(void) clear
{
    stack = [NSMutableArray array];
    incrementalImage = nil;
    [currentTool clear];
    [self setNeedsDisplay];
    [self onDrawSketch];
}

-(void) undo
{
    UIImage *prev = [stack stackPop];
    if(prev && (prev != [NSNull null])){
        incrementalImage = prev;
    }
    else{
        incrementalImage = nil;
    }
    
    [currentTool clear];
    [self setNeedsDisplay];
    [self onDrawSketch];
    
}

// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
    [incrementalImage drawInRect:rect];
    [currentTool render];
}

-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [currentTool touchesBegan:touches withEvent:event];
}

-(void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [currentTool touchesMoved:touches withEvent:event];
}

-(void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    if([currentTool touchesEnded:touches withEvent:event]){
        [self takeSnapshot];
    }
    
}

-(void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    if([currentTool touchesCancelled:touches withEvent:event]){;
        [self takeSnapshot];
    }
}

-(void)takeSnapshot
{
    [self setViewImage:[self drawBitmap]];
    [currentTool clear];
}

-(void)onDrawSketch
{
    // get container view
    SketchViewContainer * parent = (SketchViewContainer *)self.superview;
    
    if(parent){
        parent.onDrawSketch(@{
            @"stackCount": @([stack count])
        });
    }
}

-(UIImage *)drawBitmap
{
    UIGraphicsBeginImageContextWithOptions(self.bounds.size, NO, 0.0);
    [self drawRect:self.bounds];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

@end

File: android/src/main/java/com/rnsketchview/SketchView.java
package com.rnsketchview;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.view.MotionEvent;
import android.view.View;
import java.util.LinkedList;

import com.rnsketchview.tools.SketchTool;
import com.rnsketchview.tools.EraseSketchTool;
import com.rnsketchview.tools.PenSketchTool;
import com.rnsketchview.tools.RectangleTool;
import com.rnsketchview.tools.ArrowTool;
import com.rnsketchview.tools.TextTool;


public class SketchView extends View {

    SketchViewContainer mContainer;
    int maxUndo = 10;
    SketchTool currentTool;
    PenSketchTool penTool;
    EraseSketchTool eraseTool;
    RectangleTool rectangleTool;
    ArrowTool arrowTool;
    TextTool textTool;

    Bitmap incrementalImage;
    LinkedList<Bitmap> stack;

    public SketchView(Context context, SketchViewContainer container) {
        super(context);

        mContainer = container;
        stack = new LinkedList<Bitmap>();
        penTool = new PenSketchTool(this);
        eraseTool = new EraseSketchTool(this);
        rectangleTool = new RectangleTool(this);
        arrowTool = new ArrowTool(this);
        textTool = new TextTool(this);

        setToolType(SketchTool.TYPE_PEN);
        setBackgroundColor(Color.TRANSPARENT);
    }

    public void setToolType(int toolType) {

        // clear/commit current tool
        commit();

        switch (toolType) {
            case SketchTool.TYPE_PEN:
                currentTool = penTool;
                break;
            case SketchTool.TYPE_ERASE:
                currentTool = eraseTool;
                break;
            case SketchTool.TYPE_RECTANGLE:
                currentTool = rectangleTool;
                break;
            case SketchTool.TYPE_ARROW:
                currentTool = arrowTool;
                break;
            case SketchTool.TYPE_TEXT:
                currentTool = textTool;
                break;
            default:
                currentTool = penTool;
        }
    }

    void setMaxUndo(int max){
        maxUndo = max;

        synchronized(this){
            int initialSize = stack.size();

            while(stack.size() >= maxUndo){
                stack.pollFirst();
            }
            if(initialSize != stack.size()){
                mContainer.onDrawSketch(stack.size());
            }
        }
    }

    // keep pen tool as the source of truth
    // but set all that might need use color
    public void setToolColor(int toolColor) {
        penTool.setToolColor(toolColor);
        rectangleTool.setToolColor(toolColor);
        arrowTool.setToolColor(toolColor);
        textTool.setToolColor(toolColor);
    }

    public int getToolColor() {
        return penTool.getToolColor();
    }

    public void setViewImage(Bitmap bitmap) {
        synchronized(this){
            if(incrementalImage != null){
                if(stack.size() >= maxUndo){
                    stack.pollFirst();
                }
                stack.addLast(incrementalImage);
            }
            // if prev image is null,
            // add a dummy stack element so we can undo it too
            else{
                stack.addLast(null);
            }
        }
        incrementalImage = bitmap;
        invalidate();
        mContainer.onDrawSketch(stack.size());
    }

    Bitmap drawBitmap() {
        Bitmap viewBitmap = Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(viewBitmap);
        draw(canvas);
        return viewBitmap;
    }

    public void clear() {
        synchronized(this){
            // clear stack as well
            stack = new LinkedList<Bitmap>();
        }
        incrementalImage = null;
        currentTool.clear();
        invalidate();
        mContainer.onDrawSketch(stack.size());
    }

    public void undo() {
        // null is also handled
        Bitmap prev = stack.pollLast();
        incrementalImage = prev;
        currentTool.clear();
        invalidate();
        mContainer.onDrawSketch(stack.size());
    }

    public void commit() {
        if(currentTool != null){

            if(currentTool.hasData()){
                setViewImage(drawBitmap());
                currentTool.clear();
            }
            else{
                currentTool.clear();
                invalidate();
            }
        }
    }

    public void promptData() {
        if(currentTool != null){
            currentTool.promptData();
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if(incrementalImage != null)
            canvas.drawBitmap(incrementalImage, getLeft(), getTop(), null);
        if(currentTool != null)
            currentTool.render(canvas);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        boolean value = false;

        int action = event.getAction();

        switch (action) {
            case MotionEvent.ACTION_DOWN:
                value = currentTool.onTouchDown(event);
                break;
            case MotionEvent.ACTION_MOVE:
                value = currentTool.onTouchMove(event);
                break;
            case MotionEvent.ACTION_UP:
                value = currentTool.onTouchUp(event);
                break;
            case MotionEvent.ACTION_CANCEL:
                value = currentTool.onTouchCancel(event);
                break;
        }

        if(value && (action == MotionEvent.ACTION_CANCEL || action == MotionEvent.ACTION_UP)) {
            setViewImage(drawBitmap());
            currentTool.clear();
        }

        return true;
    }

}
